### ISSUE TITLE
Less strict indexes

### ISSUE THREAD

**Issue by Tokazama** [invenia/IndexedDims.jl#1]
First and foremost, thanks for getting this package started!

With regards to this code
https://github.com/invenia/IndexedDims.jl/blob/639a22f36e80efd3f22b994097017d44f60e05c5/src/IndexedDims.jl#L15

If we have this instead...
```julia
struct IndexedDimsArray{T, N, A<:AbstractArray{T, N}, Is<:NTuple{N,<:AbstractUnitRange{Int}}} <: AbstractArray{T, N}
    data::A
    indexes::Is
end
```
...then we should have support from most of most of the `AbstractArray` API because both `LinearIndices` and `CartesianIndices` can directly use the `indexes` field.

**Comment by iamed2** [comment]
In this case the AcceleratedVector is our special indexes, which could be `["foo", "bar", "baz"]` for example

**Comment by Tokazama** [comment]
My only concern is that this permanently ties use to `AcceleratedArrays` as the only way of indexing. While this may be the best option right now, it ends up relying on maintenance of an entirely different indexing system, which became problematic with `AxisArrays`. It also means that even normal indices have to go through slower cartesian indexing.

Perhaps it would be better to open a PR with a more full implementation of what I'm getting at?

**Comment by iamed2** [comment]
I agree that not requiring AcceleratedArrays could be good. But the whole point of this package is the alternate indexing system. Unless I am missing what you are specifically referring to when you say that?
